status/calendar.go

```diff
@@ -0,0 +1,166 @@
+	URL         string
+	GetInterval time.Duration
+
+	NextEvent     chan string
+	StartingEvent chan string
+
+	wgGet     sync.WaitGroup
+	stopGet   chan struct{}
+	wgTimer   sync.WaitGroup
+	stopTimer chan struct{}
+}
+
+type eventByStart []gocal.Event
+
+func (v eventByStart) Len() int           { return len(v) }
+func (v eventByStart) Swap(i, j int)      { v[i], v[j] = v[j], v[i] }
+func (v eventByStart) Less(i, j int) bool { return v[i].Start.Before(*v[j].Start) }
+
+func (c *Calendar) Start() {
+	c.NextEvent = make(chan string)
+	c.StartingEvent = make(chan string)
+	c.stopGet = make(chan struct{})
+	c.wgGet.Add(1)
+	go c.getLoop()
+}
+
+func (c *Calendar) getLoop() {
+	var sleep time.Duration
+	var lastModified string
+	defer c.wgGet.Done()
+	for {
+		log.Printf("Calendar sleeping %s", sleep)
+		select {
+		case <-c.stopGet:
+			return
+		case <-c.Clock.After(sleep):
+		}
+
+		req, err := http.NewRequest("GET", c.URL, nil)
+		if err != nil {
+			log.Panicf("NewRequest: %s", err)
+		}
+		if lastModified != "" {
+			req.Header.Set("If-Modified-Since", lastModified)
+		}
+		resp, err := c.HTTPClient.Do(req)
+		if err != nil {
+			log.Printf("Calendar get: %s", err)
+			sleep = 1 * time.Minute
+			continue
+		}
+
+		log.Printf("Calendar response %d (length %s)", resp.StatusCode, resp.Header.Get("Content-Length"))
+		switch resp.StatusCode {
+		case http.StatusOK:
+			err = c.parse(resp.Body)
+			if err == nil {
+				lastModified = resp.Header.Get("Last-Modified")
+				log.Printf("Last modified now: %s", lastModified)
+				sleep = c.GetInterval
+			} else {
+				sleep = 1 * time.Minute
+			}
+			resp.Body.Close()
+		case http.StatusNotModified:
+			sleep = c.GetInterval
+		default:
+			log.Printf("Unexpected response %d", resp.StatusCode)
+			sleep = 1 * time.Minute
+		}
+	}
+}
+
+func (c *Calendar) parse(r io.Reader) error {
+	// Don't use https://github.com/arran4/golang-ical, it doesn't handle non-standard
+	// properties (X-...) which we use in our calendar (X-EVENTMONTH:JAN2025 -
+	// https://github.com/FOULAB/foulab-dot-org-hugo/blob/91dfe26930ac1f56b10b645a4979d480f941f647/static/ical/foulab.ics#L1299)
+
+	start, end := time.Now(), time.Now().Add(30*24*time.Hour)
+	cal := gocal.NewParser(r)
+	cal.Start, cal.End = &start, &end
+	if err := cal.Parse(); err != nil {
+		log.Printf("Parse calendar error: %s", err)
+		return err
+	}
+
+	events := cal.Events
+	log.Printf("Calendar parsed %d events", len(events))
+
+	// TODO: avoid sorting events in the past
+	sort.Sort(eventByStart(events))
+
+	// TODO: add 'morning' timer ("Events tonight: ...")
+
+	if c.stopTimer != nil {
+		close(c.stopTimer)
+		c.wgTimer.Wait()
+	}
+
+	c.stopTimer = make(chan struct{})
+	c.wgTimer.Add(1)
+	go c.timerLoop(events)
+	return nil
+}
+
+func (c *Calendar) timerLoop(events []gocal.Event) {
+	defer c.wgTimer.Done()
+
+	i := 0
+	for {
+		// TODO: make this robust against system time changes
+		now := c.Clock.Now()
+
+		// TODO: free memory for past events
```

Comment:
aren't we GETing this every 60 minutes? When will there ever be events in the past?

Reply:
GET every 60 minutes: yes.

If response is `200 OK` then we call `parse` and this restarts `timerLoop`

BUT if the response is `304 Not Modified` (this is the common case, almost every time, except ~once per month), `timerLoop` is still running, and events gradually move into the past.

Technically there is "a few KB of memory" to be saved here for events in the past. But realistically, I'll probably never implement that. Just noting that I know an improvement is technically possible.

Reply:
oh, I see. 

I wouldn't write that as a TODO. Just // past events will remain in-memory until ... calendar is updated?